### PR TITLE
Weinbe58/descretize waveform

### PR DESCRIPTION
**Pull request by weinbe58**
new code to convert arbitrary waveforms into piecewise linear given constraints and tolerances. 

This will be used in the BloqadeSchema to port protocols to JSON files.

### CODE REVIEW

File: lib/BloqadeWaveforms/src/descretize.jl [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+function descretize_waveform(wf::Waveform{PiecewiseLinear{T,Interp},T}; 
+    max_value::Real=Inf, 
+    max_slope::Real=Inf, 
+    min_step::Real=0, 
+    tol::Float64 = 1.0e-3) where {T<:Real,Interp}
+
+    c = wf.f.clocks
+    v = wf.f.values
+
+    for i in 1:length(c)-1
```
Comment (Roger-luo) [inline lib/BloqadeWaveforms/src/descretize.jl:14]:
```suggestion
    @inbounds for i in 1:length(c)-1
```

File: lib/BloqadeWaveforms/src/descretize.jl [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+function descretize_waveform(wf::Waveform; 
+    max_value::Real=Inf, 
+    max_slope::Real=Inf, 
+    min_step::Real=0, 
+    tol::Float64 = 1.0e-5)
+    
+    # TODO:
+    # need to optimize stack and intervals
+
+    stack = Tuple{Float64,Float64}[(0.0,wf.duration)]
```
Comment (Roger-luo) [inline lib/BloqadeWaveforms/src/descretize.jl:98]:
```suggestion
    stack = [(0.0,wf.duration)]
```

File: lib/BloqadeWaveforms/src/descretize.jl [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    intervals = Tuple{Float64,Float64,Float64,Float64}[]
```
Comment (Roger-luo) [inline lib/BloqadeWaveforms/src/descretize.jl:99]:
```suggestion
    intervals = NTuple{4, T}[]
```

File: lib/BloqadeWaveforms/src/descretize.jl [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    
+    push!(values,v[end])
+    push!(clocks,c[end])
+
+
+    return piecewise_linear(;clocks=clocks,values=values)
+
+end
+
+function descretize_waveform(wf::Waveform;
```
Comment (Roger-luo) [inline lib/BloqadeWaveforms/src/descretize.jl:89]:
```suggestion
function descretize_waveform(wf::Waveform{F, T}; 
```

File: lib/BloqadeWaveforms/src/descretize.jl [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+    return piecewise_linear(;clocks=clocks,values=values)
+
+end
+
+function descretize_waveform(wf::Waveform; 
+    max_value::Real=Inf, 
+    max_slope::Real=Inf, 
+    min_step::Real=0, 
+    tol::Float64 = 1.0e-5)
```
Comment (Roger-luo) [inline lib/BloqadeWaveforms/src/descretize.jl:93]:
```suggestion
    tol::Float64 = 1.0e-5) where {T}
```

File: lib/BloqadeWaveforms/src/descretize.jl [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
```
Comment (Roger-luo) [inline lib/BloqadeWaveforms/src/descretize.jl:94]:
```julia
T = promote_type(T, Float64)
```

File: lib/BloqadeWaveforms/test/descretize.jl [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    try
+        descretize_waveform(wf;max_value=1.0)
+    catch e
+        if isa(e,ErrorException) && e.msg == "Waveform exceeds maximum value."
+            exception_max_value_test = true
+        end
+    end
+
+    exception_min_step_test = false
+    try
```
Comment (Roger-luo) [inline lib/BloqadeWaveforms/test/descretize.jl:20]: use `@test_throws`

**Review by Roger-luo — CHANGES_REQUESTED** [review]

**Comment by Roger-luo** [comment]
Need to merge master to get tests pass

**Review by Roger-luo — APPROVED** [review]
LGTM

**Comment by weinbe58** [comment]
What does that mean?

**Comment by Roger-luo** [comment]
sry, it means "looks good to me" (Damn it feels good that I'm not the only one keep checking the urban dictionary :P ). 

So we usually follow the [colprac](https://github.com/SciML/ColPrac) workflow in general, so if you have merge rights and your PR passed review, you should merge it by yourself (since sometimes people may change their mind and want to further modify their own PR)

**Comment by weinbe58** [comment]
I will add documentation for this function then merge the PR.

**Pull request merged**